File: 03.binary-diagnostic/py/part2.py

import fileinput

most = []
least = []

for line in fileinput.input():
    most.append(line.strip())
    least.append(line.strip())

def countCommon(values, i):
    ones = sum([2 * int(v[i]) for v in values])
    if ones < len(values):
        return ('0', '1')
    else:
        return ('1', '0')

for i in range(12):
    if len(most) != 1:
        common = countCommon(most, i)
        most = [v for v in most if v[i] == common[0]]
    if len(least) != 1:
        common = countCommon(least, i)
        least = [v for v in least if v[i] == common[1]]

print('Solution:', int(most[0], 2) * int(least[0], 2))

File: 03.binary-diagnostic/lua/part2.lua

most = {}
least = {}

for line in io.lines() do
    table.insert(most, line)
    table.insert(least, line)
end

function countCommon(t, i)
    ones = 0
    for k, v in pairs(t) do
        if v:sub(i, i) == '1' then
            ones = ones + 2
        end
    end
    if ones < #t then
        return {'0', '1'}
    else
        return {'1', '0'}
    end
end

function filter(t, i, e)
    filtered = {}
    for k, v in pairs(t) do
        if v:sub(i, i) == e then
            table.insert(filtered, v)
        end
    end
    return filtered
end

for i = 1,12,1 do
    if #most > 1 then
        common = countCommon(most, i)
        most = filter(most, i, common[1])
    end
    if #least > 1 then
        common = countCommon(least, i)
        least = filter(least, i, common[2])
    end
end

print('Result:', tonumber(most[1], 2) * tonumber(least[1], 2))

